services/gateway/endpoints/src/main/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsRoute.java

```diff
@@ -0,0 +1,281 @@
+        return path(PATH_CLOUDEVENTS, () -> // /cloudevents
+                post(() -> // POST
+                        acceptCloudEvent(ctx, dittoHeaders)
+                )
+        );
+    }
+
+    protected Route acceptCloudEvent(final RequestContext ctx, final DittoHeaders dittoHeaders) {
+
+        return extractDataBytes(payloadSource -> {
+
+            final CompletableFuture<HttpResponse> httpResponseFuture = new CompletableFuture<>();
+
+            runWithSupervisionStrategy(payloadSource
+                    // collect the binary payload
+                    .fold(ByteString.emptyByteString(), ByteString::concat)
+                    // map the payload to a cloud event
+                    .map(payload -> toCloudEvent(ctx, dittoHeaders, payload))
+                    // validate the cloud event
+                    .map(cloudEvent -> validateCloudEvent(cloudEvent, ctx, dittoHeaders))
+                    // process the event
+                    .map(cloudEvent -> {
+                        try {
+                            // DON'T replace this try-catch by .recover: The supervising strategy is called before recovery!
+                            final Optional<Signal<?>> optionalSignal = jsonToDittoSignal(cloudEvent.getData(), dittoHeaders);
+                            if (optionalSignal.isEmpty()) {
+                                return new Status.Failure(MissingPayloadException
+                                        .withDetailedInformationBuilder()
+                                        .dittoHeaders(dittoHeaders)
+                                        .build());
+                            }
+
+                            final Signal<?> signal = optionalSignal.get();
+                            final JsonSchemaVersion schemaVersion = signal.getImplementedSchemaVersion();
+                            return signal.implementsSchemaVersion(schemaVersion) ? signal
+                                    : CommandNotSupportedException.newBuilder(schemaVersion.toInt())
+                                    .dittoHeaders(dittoHeaders)
+                                    .build();
+                        } catch (final Exception e) {
+                            return new Status.Failure(e);
+                        }
+                    })
+                    .to(Sink.actorRef(createHttpPerRequestActor(ctx, httpResponseFuture),
+                            AbstractHttpRequestActor.COMPLETE_MESSAGE))
+            );
+
+            // return with future
+
+            return completeWithFuture(httpResponseFuture);
+
+        });
+
+    }
+
+    /**
+     * Convert the request and payload to a cloud event.
+     *
+     * @param ctx The request context.
+     * @param payload The binary payload, this may contain the cloud event payload, or the fully encoded cloud
+     *         event structure.
+     * @return The cloud event
+     */
+    private @Nullable
```

Comment:
please remove the `@Nullable` - this may not be `@Nullable` as `null` is not allowed as element in Akka Streams where this method is invoked in.
I also don't see at the `MessageReader.toEvent()` javadoc that `null` could be returned.